### PR TITLE
feat: auto publish after creating new provision.

### LINKED ISSUE #7148: Confirm Skill endpoints screen appears to be a dead-end (until provisioning happens)
## Describe the bug

When making a bot available as a skill (which is not already published) provisioning/publishing is triggered. UX then moves to a confirm skill endpoints screen. This isn't proceedable (you can go back) and it's not clear what I should do next. I was patient, and after provisioning/publishing completed I spotted the UX had updated with the endpoint.

Think we need a spinner/modal dialog to communicate it's waiting.

## Version

1.4.0-nightly.237101

## Browser

- [x] Electron distribution
- [ ] Chrome
- [ ] Safari
- [ ] Firefox
- [ ] Edge

## OS

- [ ] macOS
- [x] Windows
- [ ] Ubuntu

## Screenshots

image

### PR DESCRIPTION
**Pull request by alanlong9278**
## Description

Auto publish after creating new provision.

## Task Item

closes #7148

### CODE REVIEW

File: Composer/packages/client/src/pages/design/exportSkillModal/index.tsx [hunk 1]
```diff
@@ -71,6 +89,84 @@ const ExportSkillModal: React.FC<ExportSkillModalProps> = ({ onSubmit, onDismiss
...
+    }
+    const responseData = apiResponse.data;
+
+    if (responseData.status !== ApiStatus.Publishing) {
+      // Show result notifications
+      const displayedNotification = publishNotificationRef.current;
+      if (displayedNotification) {
+        const currentTarget = publishTargets?.find((target) => target.name === publishUpdater.getPublishTargetName());
+        const url = currentTarget
+          ? `https://${JSON.parse(currentTarget.configuration).hostname}.azurewebsites.net/manifests/${
```
Comment (boydc2014) [inline Composer/packages/client/src/pages/design/exportSkillModal/index.tsx:119]: does this have to be hard-coded

Comment (alanlong9278) [reply Composer/packages/client/src/pages/design/exportSkillModal/index.tsx:119]: Yep.

Comment (boydc2014) [reply Composer/packages/client/src/pages/design/exportSkillModal/index.tsx:119]: after checking with @luhan2017  and @alanlong9278 , this url is a pattern in azure that we are replying on in multiple places. It should be safe to infer endpoint based on hostname in azure environment.

**Review by boydc2014 — DISMISSED** [review]

File: Composer/packages/client/src/pages/design/exportSkillModal/index.tsx (hunk 1 above)
Comment (boydc2014) [reply Composer/packages/client/src/pages/design/exportSkillModal/index.tsx:119]: FYI, @cwhitten, i think we can keep it as for now, until we are going to support different cloud provider and in that time we will have to update more places than just this one

File: Composer/packages/client/src/pages/design/exportSkillModal/index.tsx [hunk 2]
```diff
@@ -71,6 +89,84 @@ const ExportSkillModal: React.FC<ExportSkillModalProps> = ({ onSubmit, onDismiss
...
+      stopUpdater(publishUpdater);
+      navigate(`bot/${projectId}/publish/all`);
+    }
+  };
+
+  useEffect(() => {
+    const publish = async () => {
+      if (!publishTargets || publishTargets.length === 0) return;
+      const currentTarget = publishTargets.find((item) => {
+        const config = JSON.parse(item.configuration);
```
Comment (srinaath) [inline Composer/packages/client/src/pages/design/exportSkillModal/index.tsx:135]: wrap in try catch

File: Composer/packages/client/src/pages/design/exportSkillModal/index.tsx [hunk 3]
```diff
@@ -71,6 +89,84 @@ const ExportSkillModal: React.FC<ExportSkillModalProps> = ({ onSubmit, onDismiss
...
+        return (
+          config.settings &&
+          config.settings.MicrosoftAppId &&
+          config.hostname &&
+          config.settings.MicrosoftAppId.length > 0 &&
+          config.hostname.length > 0
+        );
+      });
+      if (isCreateProfileFromSkill && currentTarget) {
+        const skillPublishPenddingNotificationCard = getSkilPendingNotificationCardProps();
```
Comment (srinaath) [inline Composer/packages/client/src/pages/design/exportSkillModal/index.tsx:145]: **nit: typo here getSkillPendingNotificationCardProps**

File: Composer/packages/client/src/pages/publish/Notifications.tsx [hunk 4]
```diff
@@ -71,7 +71,10 @@ export const getPublishedNotificationCardProps = (item: BotStatus): CardProps =>
   };
 };
 
-export const getSkillPublishedNotificationCardProps = (item: BotStatus, url?: string): CardProps => {
+export const getSkillPublishedNotificationCardProps = (
+  item: { status: number; [key: string]: any },
```
Comment (srinaath) [inline Composer/packages/client/src/pages/publish/Notifications.tsx:75]: Record\<string, any>. Can we avoid any here?

File: Composer/packages/client/src/recoilModel/dispatchers/provision.ts [hunk 5]
```diff
@@ -135,7 +135,7 @@ export const provisionDispatcher = () => {
           });
 
           // update publishTargets
-          callbackHelpers.set(settingsState(projectId), (settings) => {
+          await callbackHelpers.set(settingsState(projectId), (settings) => {
```
Comment (srinaath) [inline Composer/packages/client/src/recoilModel/dispatchers/provision.ts:138]: recoil set does not need an await. it does things synchronously.

File: Composer/packages/server/src/controllers/publisher.ts [hunk 6]
```diff
@@ -53,16 +54,23 @@ export const PublishController = {
...
+    currentPublishTarget = publishTarget;
     const currentProject = await BotProjectService.getProjectById(projectId, user);
 
     // deal with publishTargets not exist in settings
     const publishTargets = currentProject.settings?.publishTargets || [];
     const allTargets = [defaultPublishConfig, ...publishTargets];
 
     const profiles = allTargets.filter((t) => t.name === target);
-    const profile = profiles.length ? profiles[0] : undefined;
+    let profile: PublishTarget;
```
Comment (srinaath) [inline Composer/packages/server/src/controllers/publisher.ts:67]: The same lines are repeated on line 68. Wrap it in a function?

**Comment by srinaath** [comment]
@alanlong9278 Minor comments to be resolve before u can merge.

File: Composer/packages/client/src/pages/design/exportSkillModal/index.tsx (hunk 1 above)
Comment (hatpick) [reply Composer/packages/client/src/pages/design/exportSkillModal/index.tsx:119]: even if that's the case it better to be encapsulated in a function so it's easier for extension in the future.

File: Composer/packages/client/src/recoilModel/DispatcherWrapper.tsx [hunk 7]
```diff
@@ -123,7 +123,7 @@ export const DispatcherWrapper = ({ children }) => {
         if (filePersistence.isErrorHandlerEmpty()) {
           filePersistence.registerErrorHandler(setProjectError);
         }
-        filePersistence.notify(assets, previousAssets);
+        await filePersistence.notify(assets, previousAssets);
```
Comment (hatpick) [inline Composer/packages/client/src/recoilModel/DispatcherWrapper.tsx:126]: @srinaath is this change required?

File: Composer/packages/client/src/recoilModel/dispatchers/notification.ts [hunk 8]
```diff
@@ -53,10 +53,17 @@ export const notificationDispatcher = () => {
     set(notificationsState(id), (notification) => ({ ...notification, hidden: true }));
   });
 
+  const updateNotification = useRecoilCallback(
+    (callbackHelper: CallbackInterface) => (id: string, newValue: CardProps) => {
```
Comment (hatpick) [inline Composer/packages/client/src/recoilModel/dispatchers/notification.ts:57]: newValue type should be ```Partial<CardProps>``` so the called can change it partially.

change updateNotificationInternal accordingly to accept partial props changes:
```typescript
export const updateNotificationInternal = ({ set }: CallbackInterface, id: string, newValue: Partial<CardProps>) => {
// I'm not sure why id is being re-set here?
  set(notificationsState(id), (current) => { ...current, ...newValue, id: id });
  // check if notification exist
  set(notificationIdsState, (notificationIds) => {
    if (notificationIds.some((notificationId) => notificationId === id)) {
      return notificationIds;
    } else {
      return [...notificationIds, id];
    }
  });
};
```

Comment (hatpick) [reply Composer/packages/client/src/recoilModel/dispatchers/notification.ts:57]: This will allow a caller to change notification partially, for example, change the status only by passing ONLY status.

File: Composer/packages/client/src/recoilModel/dispatchers/publisher.ts [hunk 9]
```diff
@@ -190,7 +193,7 @@ export const publisherDispatcher = () => {
         // add job id to storage
         const publishJobIds = publishStorage.get('jobIds') || {};
         publishJobIds[`${projectId}-${target.name}`] = response.data.id;
-        publishStorage.set('jobIds', publishJobIds);
+        await publishStorage.set('jobIds', publishJobIds);
```
Comment (hatpick) [inline Composer/packages/client/src/recoilModel/dispatchers/publisher.ts:196]: why await?

File: Composer/packages/server/src/controllers/publisher.ts (hunk 6 above)
Comment (hatpick) [reply Composer/packages/server/src/controllers/publisher.ts:67]: +1 for @srinaath 
also you can simplify it by:
```typescript
let profile: PublishTarget = profiles[0];
if(!profile) {
  profile = publishTarget;
  currentProject.settings?.publishTargets && currentProject.settings?.publishTargets.push(publishTarget);
}
```

**Review by hatpick — CHANGES_REQUESTED** [review]

File: Composer/packages/client/src/recoilModel/DispatcherWrapper.tsx (hunk 7 above)
Comment (srinaath) [reply Composer/packages/client/src/recoilModel/DispatcherWrapper.tsx:126]: @hatpick Nope we shouldn't be requiring to await on filePersistence. This notify should be more like a fire and forget model where a task is added to the taskQueue and thigns are processed in the background.

@alanlong9278 Let's try to avoid awaiting here as we do not know the side effects of this change. In addition, why do you need to avoid?

File: Composer/packages/client/src/pages/design/exportSkillModal/index.tsx (hunk 1 above)
Comment (alanlong9278) [reply Composer/packages/client/src/pages/design/exportSkillModal/index.tsx:119]: Sure, I will wrap it into a function.

File: Composer/packages/client/src/pages/design/exportSkillModal/index.tsx (hunk 2 above)
Comment (alanlong9278) [reply Composer/packages/client/src/pages/design/exportSkillModal/index.tsx:135]: Fine

File: Composer/packages/client/src/pages/design/exportSkillModal/index.tsx (hunk 3 above)
Comment (alanlong9278) [reply Composer/packages/client/src/pages/design/exportSkillModal/index.tsx:145]: Fixed

File: Composer/packages/client/src/pages/publish/Notifications.tsx (hunk 4 above)
Comment (alanlong9278) [reply Composer/packages/client/src/pages/publish/Notifications.tsx:75]: It just should check status type, so i set any to other properties

File: Composer/packages/client/src/recoilModel/dispatchers/provision.ts (hunk 5 above)
Comment (alanlong9278) [reply Composer/packages/client/src/recoilModel/dispatchers/provision.ts:138]: remove it

File: Composer/packages/server/src/controllers/publisher.ts (hunk 6 above)
Comment (alanlong9278) [reply Composer/packages/server/src/controllers/publisher.ts:67]: Wraped in a function

File: Composer/packages/client/src/recoilModel/dispatchers/publisher.ts (hunk 9 above)
Comment (alanlong9278) [reply Composer/packages/client/src/recoilModel/dispatchers/publisher.ts:196]: removed

File: Composer/packages/client/src/pages/botProject/CreatePublishProfileDialog.tsx [hunk 10]
```diff
@@ -22,10 +22,11 @@ import { actionButton } from './styles';
 
 type CreatePublishProfileDialogProps = {
   projectId: string;
+  setIsCreateProfileFromSkill: (isCreateProfileFromSkill: boolean) => void;
```
[the remark this answers is not in the document]
Comment (alanlong9278) [reply Composer/packages/client/src/pages/botProject/CreatePublishProfileDialog.tsx:25]: Rename

File: Composer/packages/client/src/pages/design/exportSkillModal/content/SelectProfile.tsx [hunk 11]
```diff
@@ -201,7 +208,10 @@ export const SelectProfile: React.FC<ContentProps> = ({ manifest, setSkillManife
     </div>
   ) : (
     <div>
-      <CreatePublishProfileDialog projectId={projectId}></CreatePublishProfileDialog>
+      <CreatePublishProfileDialog
+        projectId={projectId}
+        setIsCreateProfileFromSkill={setIsCreateProfileFromSkill}
+      ></CreatePublishProfileDialog>
```
[the remark this answers is not in the document]
Comment (alanlong9278) [reply Composer/packages/client/src/pages/design/exportSkillModal/content/SelectProfile.tsx:214]: removed

File: Composer/packages/client/src/pages/design/exportSkillModal/index.tsx [hunk 12]
```diff
@@ -71,6 +89,84 @@ const ExportSkillModal: React.FC<ExportSkillModalProps> = ({ onSubmit, onDismiss
     !isAdaptive ? skillConfiguration?.allowedCallers : runtimeSettings?.skills?.allowedCallers ?? []
   );
 
+  const [isCreateProfileFromSkill, setIsCreateProfileFromSkill] = useState(false);
+  let publishUpdater: PublishStatusPollingUpdater;
```
[the remark this answers is not in the document]
Comment (alanlong9278) [reply Composer/packages/client/src/pages/design/exportSkillModal/index.tsx:93]: Yep, Replaced it

File: Composer/packages/client/src/recoilModel/dispatchers/notification.ts (hunk 8 above)
Comment (alanlong9278) [reply Composer/packages/client/src/recoilModel/dispatchers/notification.ts:57]: Refactor.    `set(notificationsState(id), (current) => ({ ...current, ...newValue }));` . Id should not be reset

File: Composer/packages/client/src/recoilModel/DispatcherWrapper.tsx (hunk 7 above)
Comment (alanlong9278) [reply Composer/packages/client/src/recoilModel/DispatcherWrapper.tsx:126]: Remove it. Add `await` is for test updating file content sync. We should remove it.

File: Composer/packages/client/src/pages/botProject/create-publish-profile/PublishProfileDialog.tsx [hunk 13]
```diff
@@ -194,7 +195,8 @@ export const PublishProfileDialog: React.FC<PublishProfileDialogProps> = (props)
         arm = getTokenFromCache('accessToken');
         graph = getTokenFromCache('graphToken');
       }
-      provisionToTarget(fullConfig, config.type, projectId, arm, graph, current?.item);
+      await provisionToTarget(fullConfig, config.type, projectId, arm, graph, current?.item);
+      setIsCreateProfileFromSkill?.(true);
```
[the remark this answers is not in the document]
Comment (alanlong9278) [reply Composer/packages/client/src/pages/botProject/create-publish-profile/PublishProfileDialog.tsx:199]: It is optional. Because it has value only when coming from `share as skill`

File: Composer/packages/client/src/pages/design/exportSkillModal/constants.tsx [hunk 14]
```diff
@@ -86,6 +86,7 @@ export interface ContentProps {
   setSelectedDialogs: (dialogs: any[]) => void;
   setSelectedTriggers: (selectedTriggers: any[]) => void;
   setSkillManifest: (_: Partial<SkillManifestFile>) => void;
+  setIsCreateProfileFromSkill: (isCreateProfileFromSkill: boolean) => void;
```
[the remark this answers is not in the document]
Comment (alanlong9278) [reply Composer/packages/client/src/pages/design/exportSkillModal/constants.tsx:89]: Fixed

File: Composer/packages/client/src/pages/botProject/CreatePublishProfileDialog.tsx [hunk 15]
```diff
@@ -22,10 +22,11 @@ import { actionButton } from './styles';
 
 type CreatePublishProfileDialogProps = {
   projectId: string;
+  OnUpdateIsCreateProfileFromSkill: (isCreateProfileFromSkill: boolean) => void;
```
Comment (hatpick) [inline Composer/packages/client/src/pages/botProject/CreatePublishProfileDialog.tsx:25]: ```OnUpdateIsCreateProfileFromSkill``` should be camelcase ```onUpdateIsCreateProfileFromSkill```

**Review by hatpick — COMMENTED** [review]
There are a bunch of lint issues that should be resolved.

**Review by cwhitten — APPROVED** [review]
